### PR TITLE
Dev/multitask - STAMP extended for regression and survival analysis (v2.4 update)

### PR DESCRIPTION
**Pull request by mducducd**
Stamp v2.4
- Add support for regression and survival analysis tasks.
   - Regression with MAE loss.
   - Cox-based survival analysis, currently supporting title-level training.
- Add TransMIL (NeurIPS 2021) as a new attention-based model.
- Training now supports optional backbones: Attention (ViT), TransMIL, and MLP.
- All backbone types can perform tile-level or patient-level modeling by automatic ModelClass initialization based on feature type and task.
- Add new statistics: Regression → R², Pearson r, Pearson p, MAE, RMSE. Survival → Kaplan–Meier curve, log-rank p-value, C-index.
- Added probability heatmaps for regression and survival analysis.

### CODE REVIEW

File: src/stamp/statistics/survival.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    return int(((t_i < t_j) & (e_i == 1)).sum())
+
+
+def _cindex_auto(
+    time: np.ndarray,
+    event: np.ndarray,
+    risk: np.ndarray,
+) -> tuple[float, str, float, float, int]:
+    """Compute C-index and choose orientation (risk or -risk)."""
+    c_pos = concordance_index(time, risk, event)
```
Comment (s1787956) [inline src/stamp/statistics/survival.py]: lifelines expects survival times, not risks; therefore the outputs of the cox models must be multiplied by -1.

File: src/stamp/modeling/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+        super().__init__(time_label=time_label, status_label=status_label, **kwargs)
+        self.hparams.update({"task": "survival"})
+        self.method = method
+        self.time_label = time_label
+        self.status_label = status_label
+        # storage for validation accumulation
+        self._val_scores, self._val_times, self._val_events = [], [], []
+
+    @staticmethod
+    def cox_loss(
```
Comment (AsierRabasco) [inline src/stamp/modeling/models/__init__.py]: The loss is mostly fine, but could get a bit slow for larger N due to the risk mask. TorchSurv has an optimised implementation with the reordering of times + cumulative sums if interested

File: getting-started.md [hunk 3]
```diff
@@ -471,3 +471,44 @@ heatmaps:
...
+
+Advanced experiment settings can be specified under the `advanced_config` section in your configuration file.
+This section lets you control global training parameters, model type, and the target task (classification, regression, or survival).
+
+```yaml
+# stamp-test-experiment/config.yaml
+
+advanced_config:
+  seed: 42
+  task: "classification" # or regression/survial
```
[the remark this answers is not in the document]
Comment (mducducd) [reply getting-started.md]: Yes, the task will be part of the basic configuration, except for preprocess and heatmap (heatmap function can automatically infer the task from the model weights)

File: src/stamp/modeling/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+        # risk set mask: j is at risk for event i if T_j >= T_i
+        # (use >= per standard Cox; vectorized broadcast)
+        risk_mask = t_event[:, None] <= times[None, :]  # (R, N)
+
+        # log-sum-exp over risk sets for numerical stability
+        # log sum_j exp(score_j) for each event i
+        max_scores = scores.max()  # stability
+        lse = (
+            torch.log((risk_mask * torch.exp(scores - max_scores)).sum(dim=1))
+            + max_scores
```
[the remark this answers is not in the document]
Comment (mducducd) [reply src/stamp/modeling/models/__init__.py]: Currently, this one has the best results from my experiments.
I will try to remove it and retrain the model, then tell you if the performance is more or less

**Comment by mducducd** [comment]
I’ve added both slide-level and patient-level support for regression and survival tasks (only mlp can be trained). Just a few more steps are needed to add tests.

**Review by s1787956 — COMMENTED** [review]
I am very happy, great job Duc! I think we can merge tomorrow.

File: src/stamp/modeling/train.py [hunk 5]
```diff
@@ -243,77 +290,68 @@ def setup_dataloaders_for_training(
     Returns:
         train_dl, valid_dl, categories, feature_dim, train_patients, valid_patients
     """
-    # Sample count for training
-    log_total_class_summary(patient_to_data, categories)
 
     # Stratified split
```
Comment (AsierRabasco) [inline src/stamp/modeling/train.py]: The stratified split is also needed for survival, the variable for the StratifiedKFold should be the event marker. I did not see it in the code, make sure that the stratification variable for survival is set

File: src/stamp/statistics/survival.py (hunk 1 above)
Comment (AsierRabasco) [reply src/stamp/statistics/survival.py]: There is no orientation to pick. If you are using the lifelines c-index, it expects the scores to be bigger if you survive longer, but with the cox loss we get bigger scores if we survive for shorter times. Just multiply the scores that come out of the network with -1  when you pass them to the c-index function and it will be correct to what lifelines expects. 

With your method, I fear that if we have a fold that misinterprets the risk (image c-Index of 0.10, incorrect ordering but very consistent), we will not know due to the method checking for both options, so we will not distinguish between a model that is consistently correct and one that is consistently incorrect. You could end up with KM that are inverted for example. This is very important to get right
